tasty-query/shared/src/main/scala/tastyquery/Classpaths.scala

```diff
@@ -26,35 +26,74 @@ object Classpaths:
     * [[Contexts.Context]]. The latter gives semantic access to all the
     * definitions on the classpath.
     */
-  final class Classpath private (val packages: IArray[PackageData]):
+  final class Classpath private (val entries: IArray[Classpath.Entry]):
+
+    /** Returns the concatenation of this classpath with `other`.
+      * This is useful for structural sharing of [[Classpath.Entry Classpath Entries]]. e.g. in the following example
+      * the standard library is loaded once and shared between two classpaths:
+      * ```scala
+      * val stdLibCp = ClasspathLoaders.read(standardLibraryPaths)
+      * val libV101Cp = ClasspathLoaders.read(List(Paths.get("path/to/lib-1.0.1.jar"))) ++ stdLibCp
+      * val libV102Cp = ClasspathLoaders.read(List(Paths.get("path/to/lib-1.0.2.jar"))) ++ stdLibCp
+      * ```
+      */
+    def ++(other: Classpath): Classpath = Classpath(entries ++ other.entries)
+
+    /** Returns a view of all the packages in this [[Classpath]], may contain duplicate package names when
+      * the same package is present in multiple classpath entries.
+      */
+    def packages: Iterable[PackageData] = entries.view.flatMap(_.packages)
+
+    /** Filter a classpath so it only contains roots that match the given binary names. */
     def withFilter(binaryNames: List[String]): Classpath =
-      def packageAndClass(binaryName: String): (String, String) = {
+
+      def packageAndClass(binaryName: String): (String, String) =
         val lastSep = binaryName.lastIndexOf('.')
         if lastSep == -1 then ("", binaryName)
-        else {
+        else
           import scala.language.unsafeNulls
           val packageName = binaryName.substring(0, lastSep)
           val className = binaryName.substring(lastSep + 1)
           (packageName, className)
+
+      def filterEntry(entry: Classpath.Entry, lookup: Map[String, List[String]]) =
+        val packages = entry.packages.collect {
+          case pkg if lookup.contains(pkg.dotSeparatedName) =>
+            val tastys = pkg.tastys.filter(t => lookup(pkg.dotSeparatedName).contains(t.binaryName))
+            val classes = pkg.classes.filter(c => lookup(pkg.dotSeparatedName).contains(c.binaryName))
+            if tastys.isEmpty && classes.isEmpty then None // perhaps the class came from another entry
+            else Some(PackageData(pkg.dotSeparatedName, classes, tastys))
```

Comment:
It shouldn't be an issue to keep a `PackageData` with `classes.isEmpty && tastys.isEmpty`. So perhaps remove the test and then remove the `.flatten`.